### PR TITLE
FIX reduce method of unbound cython method

### LINKED ISSUE #2972: Cython-generated unbound class methods cannot be pickled
Building the following module (`src/module.py`)

```python
class MyClass:

    def __init__(self, a):
        self.a = a

    def my_method(self, b):
        return self.a + b
```
Using cython with the following setup script:
```python
from distutils.core import setup
from Cython.Build import cythonize

setup(ext_modules=cythonize("src/module.py"))
```

Generates a class (`MyClass`), whose unbound methods cannot be pickled:
```python
In [1]: from module import MyClass                                                                                                                                                                                 
                                                                                                                                                                                                                   
In [2]: import pickle                                                                                                                                                                                              
                                                                                                                                                                                                                   
In [3]: pickle.dumps(MyClass.my_method)                                                                                                                                                              
---------------------------------------------------------------------------                                                                                                                                        
PicklingError                             Traceback (most recent call last)                                                                                                                                        
<ipython-input-3-cf09bde7c7ba> in <module>                                                                                                                                                                         
----> 1 pickle.dumps(MyClass.my_method)                                                                                                                                                              
                                                                                                                                                                                                                   
PicklingError: Can't pickle <cyfunction MyClass.my_method at 0x7f8d130e9ea8>: attribute lookup
my_method on module failed
```

### PR DESCRIPTION
**Pull request by pierreglaser**
Closes #2972 
This was spotted thanks to cloudpipe/cloudpickle#259

### CODE REVIEW

**Comment by pierreglaser** [comment]
OK @scoder I think this PR is ready for review :)

**Comment by pierreglaser** [comment]
The test is skipped for early python version because for python versions `<= 3.4`, pickling unbound methods also fails for classic class. This bugfix simply fixes a regression.

**Comment by scoder** [comment]
Hmmm … I guess crossing the Py2/3 boundary with pickles isn't (well) supported anyway, so it's probably ok to make this incompatible. Does this also impact bound methods? Is there a test for those already?

**Comment by pierreglaser** [comment]
Bound methods are `method` instances (as opposed to unbound methods, which are `cyfunction` instances), and  the `method` class implements a `__reduce__` method already. But it does not harm to add a line checking this in the doctest.

File: tests/run/reduce_pickle.pyx [hunk 1]
```diff
@@ -300,5 +300,5 @@ if sys.version_info[:2] >= (3, 5):
         True
         >>> bound_method = pickle.loads(pickle.dumps(MyClass().my_method))
```
Comment (scoder) [inline tests/run/reduce_pickle.pyx]: And now that the method takes an argument, you could even call `bound_method(42)` here and see if it does the right thing.

Comment (pierreglaser) [reply tests/run/reduce_pickle.pyx]: Yes. I encountered an obscure failure in the CI that I cannot reproduce locally after https://github.com/cython/cython/pull/2969/commits/1e2c7ee88fc6eb9b386d08642b1fdb6c81e997a5, so I am enriching the tests very incrementally to figure out what could have cause the failure. As soon as this turns green I'll add an assertion on the `method`'s result.

Comment (pierreglaser) [reply tests/run/reduce_pickle.pyx]: It's looking good now, so I'll add the assertion.

**Comment by scoder** [comment]
Thanks!

**Comment by jakirkham** [comment]
Would it be possible to backport this to 0.29? Happy to submit a PR if that would help :)

**Comment by scoder** [comment]
Seems reasonable. I've picked it over.

**Comment by jakirkham** [comment]
Great thank you Stefan! 😄

**Pull request merged**